File: packages/typescript/src/rules/prefer-single-sentence-comment.ts
/**
 * @fileoverview prefer-single-sentence-comment — two sentences should be reduced to one.
 * Examples: https://github.com/sarj-ai/standards/blob/main/packages/typescript/tests/rules/prefer-single-sentence-comment.test.ts
 */

import { createRule } from "./_docs.js";
import { proseGroups, sentenceUnits } from "./_prose-budget.js";

type MessageIds = "preferOneSentence";
type Options = readonly [];

export default createRule<Options, MessageIds>({
  name: "prefer-single-sentence-comment",
  meta: {
    type: "suggestion",
    docs: { description: "Prefer one sentence and self-documenting code over a two-sentence comment." },
    schema: [],
    messages: {
      preferOneSentence: "Two-sentence comment — prefer one sentence and self-documenting code.",
    },
  },
  defaultOptions: [],
  create(context) {
    return {
      Program(): void {
        for (const group of proseGroups(context.filename, context.sourceCode)) {
          if (!group.hasTypedTags && sentenceUnits(group.text) === 2) {
            context.report({ node: group.comment, messageId: "preferOneSentence" });
          }
        }
      },
    };
  },
});

File: packages/python/src/sarj_python_lint/rules/prefer_single_sentence_comment.py
"""SARJ090 — Two sentences of comment prose should become one.

Examples: https://github.com/sarj-ai/standards/blob/main/packages/python/tests/rules/test_prefer_single_sentence_comment.py
"""

from __future__ import annotations

from typing import TYPE_CHECKING, final, override

from sarj_python_lint.rule_base import Diagnostic, Rule, Severity
from sarj_python_lint.rules._prose_budget import groups, sentence_units


if TYPE_CHECKING:
    from pathlib import Path


@final
class PreferSingleSentenceComment(Rule):
    _WARNING_SENTENCES = 2
    id = "prefer-single-sentence-comment"
    code = "SARJ090"
    description = "Two-sentence comment — prefer one sentence and self-documenting code."

    @override
    def check(self, path: Path, source: str) -> list[Diagnostic]:
        return [
            Diagnostic(path, group.line, group.col, self.code, self.description, Severity.WARNING)
            for group in groups(path, source)
            if not group.typed_sections and sentence_units(group.text) == self._WARNING_SENTENCES
        ]
